Build runtime and libraries for WebAssembly

morganbr opened the pull request:
Changes to make the runtime and libraries build for WebAssembly. That's enough to make "build wasm" succeed (although it doesn't run tests due to test script issues). Most changes are stubbing in ifdefs in architecture-sensitive code such as stack unwinding and call conversion. This also includes stub methods for the assembly stubs (using C++ as an "assembly" language since we don't have a WASM assembler). Managed libraries mostly needed to be adjusted to treat WASM like Unix.

Fixes #4505, fixes #4506, fixes #4507

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs:
```diff
@@ -652,7 +652,7 @@ internal static uint RhGetLoadedModules(TypeManagerHandle[] resultArray)
         internal static extern void RhGetCurrentThreadStackBounds(out IntPtr pStackLow, out IntPtr pStackHigh);
 
-#if PLATFORM_UNIX
+#if PLATFORM_UNIX || WASM
```
This should not be needed - PLATFORM_UNIX is defined for WASM

jkotas commented on dir.props:
```diff
@@ -241,6 +241,9 @@
     <DefineConstants>ARM64;BIT64;$(DefineConstants)</DefineConstants>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Platform)' == 'wasm'">
+    <DefineConstants>WASM;BIT32;PLATFORM_UNIX;$(DefineConstants)</DefineConstants>
```
Treating WASM as Unix flavor makes sense. We should do it consistently, e.g. set `TargetsUnix` for WASM as well.

morganbr replied:
That seems to break nuget restore (at least while building on Windows). I agree the extra msbuild definitions aren't ideal, but is it ok to keep them until we have wasm nuget packages?

jkotas replied:
If it breaks nuget restore, make a specific workaround for WASM nuget restore. I think it should be in src\Framework\Framework-native.depproj - do not pull in `Microsoft.NETCore.Native` for WASM.

morganbr replied:
Done. What are we missing?

MichalStrehovsky replied:
> What are we missing by not pulling that in?

Nothing really, because there's no package for WASM anyway. It's all of this: https://github.com/dotnet/corefx/tree/c280881a048ee9d9fbfc629ca55d755d3e2b045d/src/Native/Unix/System.Native

The native library supports CoreFX managed assemblies that need to interact with the OS.

jkotas commented on src/Native/Runtime/CMakeLists.txt:
```diff
@@ -114,10 +118,15 @@ else()
     unix/HardwareExceptions.cpp
     unix/UnixContext.cpp
     unix/UnixNativeCodeManager.cpp
-    ../libunwind/src/Unwind-EHABI.cpp
```
This change should not be needed. We should not be building Full runtime flavor for WASM. We should be building Portable flavor only.

morganbr replied:
Done.

morganbr replied:
Removed

jkotas commented on src/Native/Runtime/CMakeLists.txt:
```diff
@@ -135,12 +144,16 @@ else()
     set(ARCH_SOURCES_DIR i386)
   elseif(CLR_CMAKE_PLATFORM_ARCH_WASM)
     set(ARCH_SOURCES_DIR wasm)
+# Using C++ as our "assembly" language for WASM until we figure out how to compile WAST
```
Same here - portable runtime flavor does not include any asm files already.

morganbr replied:
Removed

morganbr replied:
Removed

jkotas commented on src/Native/Runtime/CachedInterfaceDispatch.cpp:
```diff
@@ -86,7 +86,7 @@ static void * UpdatePointerPairAtomically(void * pPairLocation,
                                           bool fFailOnNonNull)
 {
-#if defined(_X86_) || defined(_ARM_)
+#if defined(_X86_) || defined(_ARM_) || defined(_WASM_)
```
Should this ifdef rather be changed to BIT32 ?

morganbr replied:
It looks like BIT32 is managed-only

jkotas replied:
It is meant to be defined for native on 32-bit platforms as well: https://github.com/dotnet/corert/blob/master/src/Native/CMakeLists.txt#L32 . It is just does not seem to be used anywhere yet.

Or you can flip this around and use BIT64 here. There is a lot of uses of BIT64.

morganbr replied:
It looks like BIT32 is only defined for ARM, so I used !BIT64 here.

jkotas commented on src/Native/Runtime/CommonMacros.h:
```diff
@@ -10,6 +10,7 @@
 #elif defined(_TARGET_X86_)
 #elif defined(_TARGET_ARM_)
```
I would delete this ifdef. These kind of ifdefs just cause busy work when porting to new architecture.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/RHCodeMan.cpp:
```diff
@@ -189,6 +189,31 @@ void ReportRegisterSet(UInt8 regSet, REGDISPLAY * pContext, GCEnumContext * hCal
     if (regSet & CSR_MASK_X28) { ReportObject(hCallback, GetRegObjectAddr<CSR_NUM_X28>(pContext), 0); }
 }
 
+#elif defined(_TARGET_WASM_)
```
Nothing in this file is applicable to Portable flavor of the runtime. It would be better to exclude it from the portable build instead of adding ifdefs to it. (If some of the methods are require to make things compile, add NYI stubs to portable.cpp.)

morganbr replied:
Done

morganbr replied:
On further inspection, removing this file from portable is a mess since EEMethodInfo and EECodeManager are in here and there are many dependencies on those in the Module class. Is it ok to just keep this file in the portable runtime? If so, this file depends on gcdump, so I'd need to keep that as well.

jkotas replied:
I have submitted PR to compile this for ProjectN only: https://github.com/dotnet/corert/pull/4924

morganbr replied:
Thanks! I've removed changes here and in gcdump.

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -249,6 +249,10 @@ void StackFrameIterator::InternalInit(Thread * pThreadToWalk, PTR_PInvokeTransit
     }
 
+#elif defined(_TARGET_WASM_)
+    UNREFERENCED_PARAMETER(pPreservedRegsCursor);
```
These changes should not be needed once we switch to portable flavor of the runtime

morganbr replied:
Removed this one, but the rest of the changes in this file aren't under portable ifdefs so I had to keep them.

jkotas commented on src/Native/Runtime/gcdump.cpp:
```diff
@@ -256,6 +256,8 @@ void GCDump::DumpCallsiteString(UInt32 callsiteOffset, PTR_UInt8 pbCallsiteStrin
                 if (b & CSR_MASK_X26) { gcPrintf("X26 "); count++; }
                 if (b & CSR_MASK_X27) { gcPrintf("X27 "); count++; }
                 if (b & CSR_MASK_X28) { gcPrintf("X28 "); count++; }
+#elif defined (_TARGET_WASM_)
```
Include this file for the portable version of the runtime only instead.

morganbr replied:
Done (assuming you meant to put it in the full runtime, not the portable one)

morganbr replied:
Done

jkotas commented on src/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -206,14 +206,14 @@
     <Compile Include="System\Globalization\GlobalizationMode.cs" />
     <Compile Include="System\Globalization\GlobalizationMode.Windows.cs" Condition="'$(TargetsWindows)'=='true'" />
-    <Compile Include="System\Globalization\GlobalizationMode.Unix.cs" Condition="'$(TargetsUnix)'=='true'" />
+    <Compile Include="System\Globalization\GlobalizationMode.Unix.cs" Condition="'$(TargetsUnix)'=='true' or '$(Platform)' == 'wasm'" />
```
This should not be needed once we start treating wasm as Unix consistently.

morganbr replied:
Removed

jkotas commented on src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/ConstrainedCallSupport.cs:
```diff
@@ -3,11 +3,12 @@
-// The function signature of the following architectures is know for the constrained call helper functions
+// The function signature of the following architectures is known for the constrained call helper functions
 #if ARM
 #elif X86
```
I would delete this ifdef instead.

morganbr replied:
Done

jkotas commented:
Could you please also merge master into your change and resolve the conflicts?

jkotas commented on src/Native/Runtime/CMakeLists.txt:
```diff
@@ -114,8 +118,8 @@ else()
     unix/UnixNativeCodeManager.cpp
-    ../libunwind/src/Unwind-EHABI.cpp
-    ../libunwind/src/libunwind.cpp
+      ../libunwind/src/Unwind-EHABI.cpp
```
Nit: undo white space change

morganbr replied:
Fixed

morganbr commented:
@jkotas , I've rebased as well as addressing the other comments. I tried enabling tests, but it looks like msbuild doesn't pass any reference assemblies to CSC for compiling the tests if the Platform is WASM.

MichalStrehovsky commented on src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems:
```diff
@@ -648,7 +648,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\IO\Win32Marshal.cs" />
   </ItemGroup>
-  <ItemGroup Condition="$(TargetsUnix)">
+  <ItemGroup Condition="$(TargetsUnix) or '$(Platform)'=='wasm'">
```
Do we still need this?

morganbr replied:
Removed

MichalStrehovsky commented on src/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeWaitHandle.cs:
```diff
@@ -36,7 +36,7 @@ public SafeWaitHandle(IntPtr existingHandle, bool ownsHandle) : base(ownsHandle)
         protected override bool ReleaseHandle()
         {
-#if PLATFORM_UNIX
+#if PLATFORM_UNIX || WASM
```
Do we still need this?

morganbr replied:
Removed

MichalStrehovsky commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -56,7 +56,7 @@ public static void Copy(IntPtr source, byte[] destination, int startIndex, int l
         }
 
-#if PLATFORM_UNIX
+#if PLATFORM_UNIX || WASM
```
Do we still need this?

morganbr replied:
Removed

jkotas commented on src/Native/Runtime/CMakeLists.txt:
```diff
@@ -139,6 +143,7 @@ else()
     set(ARCH_SOURCES_DIR wasm)
   endif()
 
+
```
Nit: extra line

morganbr replied:
Removed

jkotas commented on src/Native/Runtime/CMakeLists.txt:
```diff
@@ -216,8 +221,8 @@ convert_to_absolute_path(PORTABLE_RUNTIME_SOURCES ${PORTABLE_RUNTIME_SOURCES})
-  add_subdirectory(Full)
-endif(NOT CLR_CMAKE_PLATFORM_WASM)
+if(NOT CLR_CMAKE_PLATFORM_ARCH_WASM)
+    add_subdirectory(Full)
```
Nit: This file seems to be using two space indentation

morganbr replied:
Fixed

jkotas commented on src/Native/Runtime/CachedInterfaceDispatch.cpp:
```diff
@@ -86,7 +86,7 @@ static void * UpdatePointerPairAtomically(void * pPairLocation,
                                           bool fFailOnNonNull)
 {
-#if defined(_X86_) || defined(_ARM_)
+#if !defined BIT64
```
The unwritten convention is to write this as: `#ifndef BIT64` or `#if !defined(BIT64)`

morganbr replied:
Added parens to match the rest of this file.

morganbr replied:
Changed to positive and used parens

jkotas commented on src/Native/Runtime/CachedInterfaceDispatch.cpp:
```diff
@@ -86,7 +86,7 @@ static void * UpdatePointerPairAtomically(void * pPairLocation,
                                           bool fFailOnNonNull)
 {
-#if defined(_X86_) || defined(_ARM_)
+#if !defined BIT64
```
I would swap the order of if and else blocks, and make the condition positive. Positive conditions are easier to read.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/EHHelpers.cpp:
```diff
@@ -301,7 +303,7 @@ EXTERN_C Int32 RhpPInvokeExceptionGuard()
 #endif
 
-#if defined(_AMD64_) || defined(_ARM_) || defined(_X86_) || defined(_ARM64_)
+#if defined(_AMD64_) || defined(_ARM_) || defined(_X86_) || defined(_ARM64_) || defined(_WASM_)
```
If this ifdef serving any purpose since you do not seem to be able to use it for not-yet-implemented WASM EH? If it is not useful, delete it instead.

morganbr replied:
This one is a little bit useful since it provides an empty implementation for RhpThrow* for new architectures. Since we're using the portable runtime, the bootstrapper is our implementation for wasm at the moment.

jkotas commented on src/Native/Runtime/EHHelpers.cpp:
```diff
@@ -412,6 +414,9 @@ static UIntNative UnwindWriteBarrierToCaller(
     pContext->SetSp(sp+sizeof(UIntNative)); // pop the stack
 #elif defined(_ARM_) || defined(_ARM64_)
     UIntNative adjustedFaultingIP = pContext->GetLr();
+#elif defined(_WASM_)
```
I would change this to:

```
#else
    UIntNative adjustedFaultingIP = 0;
    PORTABILITY_ASSERT("UnwindWriteBarrierToCaller");
#endif
```

so that it compiles fine for new ports.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/MiscHelpers.cpp:
```diff
@@ -599,6 +602,8 @@ COOP_PINVOKE_HELPER(UInt8 *, RhGetJmpStubCodeTarget, (UInt8 * pCodeOrg))
 #elif _TARGET_ARM64_
         PORTABILITY_ASSERT("@TODO: FIXME:ARM64");
+#elif _TARGET_WASM_
+        PORTABILITY_ASSERT("@TODO: FIXME:WASM");
```
Change this to:

```
#else
    PORTABILITY_ASSERT("RhGetJmpStubCodeTarget");
#endif
```

morganbr replied:
Done.

jkotas commented on src/Native/Runtime/Portable/CMakeLists.txt:
```diff
@@ -11,6 +11,26 @@ add_definitions(-DUSE_PORTABLE_HELPERS)
 
 add_library(PortableRuntime STATIC ${COMMON_RUNTIME_SOURCES} ${PORTABLE_RUNTIME_SOURCES})
 
```
You may need `get_compile_definitions(DEFINITIONS)` here just like in the Full version so that the pre-processor gets the right set of defines.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/RHCodeMan.cpp:
```diff
@@ -933,7 +932,6 @@ UIntNative EECodeManager::GetConservativeUpperBoundForOutgoingArgs(GCInfoHeader
             // value by an offset that is recorded in the info header.  Recover the address of the
             // pushed RBP value by subtracting this offset.
             upperBound = pContext->GetFP() - pInfoHeader->GetFramePointerOffset();
-
```
Nit: undo white space change

morganbr replied:
Done

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -458,6 +458,10 @@ void StackFrameIterator::InternalInit(Thread * pThreadToWalk, PTR_PAL_LIMITED_CO
     m_RegDisplay.pR10 = NULL;
     m_RegDisplay.pR11 = NULL;
+
+#elif defined(_TARGET_WASM_)
```
It would be better to change this to `#elif defined(_TARGET_X86_) || defined(_TARGET_AMD64_)` and move the `PORTABILITY_ASSERT` into the `#else` block.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/wasm/AsmOffsetsCpu.h:
```diff
@@ -0,0 +1,36 @@
+PLAT_ASM_OFFSET(14, ExInfo, m_frameIter)
+PLAT_ASM_OFFSET(9C, ExInfo, m_notifyDebuggerSP)
+
+PLAT_ASM_OFFSET(0, PInvokeTransitionFrame, m_RIP)
```
PInvokeTransitionFrame should not be needed for WASM here.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/inc/rhbinder.h:
```diff
@@ -749,6 +749,8 @@ struct PInvokeTransitionFrame
 #elif defined(_TARGET_ARM_)
 #define OFFSETOF__Thread__m_pTransitionFrame 0x2c
+#elif defined(_TARGET_WASM_)
+#define OFFSETOF__Thread__m_pTransitionFrame 0x2c
```
This should not be needed.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/inc/gcinfo.h:
```diff
@@ -302,6 +302,49 @@ enum ScratchRegMask
     SR_MASK_LR = 0x40000,
 };
 
+#elif defined (_TARGET_WASM_)
```
The changes in this file should not be needed anymore.

morganbr replied:
Removed.

jkotas commented on src/Native/Runtime/PalRedhawkCommon.h:
```diff
@@ -112,6 +112,15 @@ struct PAL_LIMITED_CONTEXT
     void SetIp(UIntNative ip) { IP = ip; }
     void SetSp(UIntNative sp) { Rsp = sp; }
     UIntNative GetFp() const { return Rbp; }
+#elif defined _TARGET_WASM_
```
Change this to `#elif defined(_TARGET_X86_) || defined(_TARGET_AMD64_)` and move the dummy implementation with `PORTABILITY_ASSERTs` into the #else block.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -614,6 +618,10 @@ void StackFrameIterator::UpdateFromExceptionDispatch(PTR_StackFrameIterator pSou
     m_RegDisplay.pR14 = thisFuncletPtrs.pR14;
     m_RegDisplay.pR15 = thisFuncletPtrs.pR15;
+
+#elif defined(_TARGET_WASM_)
```
Change this to `#elif defined(_TARGET_X86_) || defined(_TARGET_AMD64_)` and move the `PORTABILITY_ASSERT` into the `#else` block.

morganbr replied:
Done

morganbr commented on src/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -119,6 +119,7 @@ See the LICENSE file in the project root for more information.
       <IlcArg Condition="$(DebugSymbols) == 'true'" Include="-g" />
       <IlcArg Condition="$(IlcGenerateMapFile) == 'true'" Include="--map:$(NativeIntermediateOutputPath)%(ManagedBinary.Filename).map.xml" />
       <IlcArg Condition="$(RdXmlFile) != ''" Include="--rdxml:$(RdXmlFile)" />
+      <ILcArg Condition="'$(Platform)' == 'wasm'" Include="--wasm" />
```
It worked a bit better with the test system while we don't have full support. I've added support for the flag, but I'll keep the --wasm flag until we have the whole test system able to handle the concept of a wasm architecture.

morganbr commented:
@MichalStrehovsky, @jkotas, I've updated with comments and also fixed the test system so that the wasm build runs the wasm testing (in the same way that the other builds do, no new test functionality). If that looks ok to you, I'll get it rebased (the conflict in targets is a trivial merge).

morganbr commented:
Due to the myget purge, I did end up rebasing.

jkotas commented on src/Native/Runtime/MiscHelpers.cpp:
```diff
@@ -539,6 +539,9 @@ COOP_PINVOKE_HELPER(UInt8 *, RhGetCodeTarget, (UInt8 * pCodeOrg))
 #elif _TARGET_ARM64_
         UNREFERENCED_PARAMETER(unboxingStub);
         PORTABILITY_ASSERT("@TODO: FIXME:ARM64");
+#elif _TARGET_WASM_
```
Nit: Put this into `#else` instead of `#error Unsupported architecture`.

morganbr replied:
Done

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -489,6 +490,8 @@ void StackFrameIterator::InternalInit(Thread * pThreadToWalk, PTR_PAL_LIMITED_CO
     m_RegDisplay.pR10 = NULL;
     m_RegDisplay.pR11 = NULL;
 #endif // _TARGET_AMD64_
+#else // _TARGET_ARM
```
Nit: Delete `// _TARGET_ARM` comment

morganbr replied:
Done

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -626,6 +630,8 @@ void StackFrameIterator::UpdateFromExceptionDispatch(PTR_StackFrameIterator pSou
 #endif // _TARGET_AMD64_
+#else
+    PORTABILITY_ASSERT("StackFrameIterator::UpdateFromExceptionDispatch");
 #endif // _TARGET_ARM_
```
Nit: Delete `// _TARGET_ARM` comment

morganbr replied:
Done

jkotas commented on src/Native/Runtime/StackFrameIterator.cpp:
```diff
@@ -1103,6 +1130,10 @@ void StackFrameIterator::UnwindCallDescrThunk()
     // the CallDescrFunction
     newSP += sizeof(CALL_DESCR_CONTEXT) - offsetof(CALL_DESCR_CONTEXT, Rbp);
+
+#elif defined(_TARGET_WASM_)
```
Make this `#else`

morganbr replied:
Done

jkotas commented on src/Native/Runtime/ThunksMapping.cpp:
```diff
@@ -202,6 +205,11 @@ EXTERN_C REDHAWK_API void* __cdecl RhAllocateThunksMapping()
             UNREFERENCED_PARAMETER(pCurrentThunkAddress);
             /* TODO */ ASSERT_UNCONDITIONALLY("NYI");
+
+#elif _TARGET_WASM_
```
Make this `#else`

morganbr replied:
Done

morganbr commented:
Thanks for all the help, @jkotas and @MichalStrehovsky!
